File: packages/protocol/test/enums.test.ts
import { readdirSync, readFileSync } from "node:fs";
import { describe, expect, it } from "vitest";
import * as protocol from "../src";
import { GOLDEN_DIR, golden } from "./golden";

/** Every readonly array the generator added, by the name of its type. */
const ARRAYS: Record<string, readonly string[]> = {
  ActivityKind: protocol.ActivityKindValues,
  ActivityState: protocol.ActivityStateValues,
  AgentKind: protocol.AgentKindValues,
  AgentStatus: protocol.AgentStatusValues,
  CIState: protocol.CIStateValues,
  CardState: protocol.CardStateValues,
  CardViewMode: protocol.CardViewModeValues,
  ChatApprovalState: protocol.ChatApprovalStateValues,
  ChatMessageKind: protocol.ChatMessageKindValues,
  ChatPlanState: protocol.ChatPlanStateValues,
  ChatRefusalReason: protocol.ChatRefusalReasonValues,
  ChatTargetKind: protocol.ChatTargetKindValues,
  DeviceKind: protocol.DeviceKindValues,
  DiffFileStatus: protocol.DiffFileStatusValues,
  DiffLineKind: protocol.DiffLineKindValues,
  ErrorCode: protocol.ErrorCodeValues,
  EventType: protocol.EventTypeValues,
  FilterKey: protocol.FilterKeyValues,
  FeedKind: protocol.FeedKindValues,
  FrameType: protocol.FrameTypeValues,
  HoldRefusalReason: protocol.HoldRefusalReasonValues,
  LabelColor: protocol.LabelColorValues,
  MoveRefusalReason: protocol.MoveRefusalReasonValues,
  NeedsReasonKind: protocol.NeedsReasonKindValues,
  NoticeKind: protocol.NoticeKindValues,
  PermissionMode: protocol.PermissionModeValues,
  ProgressStatus: protocol.ProgressStatusValues,
  ProjectSource: protocol.ProjectSourceValues,
  ProjectView: protocol.ProjectViewValues,
  ResyncReason: protocol.ResyncReasonValues,
  SessionState: protocol.SessionStateValues,
  SortDirection: protocol.SortDirectionValues,
  Swimlane: protocol.SwimlaneValues,
  TerminalKey: protocol.TerminalKeyValues,
  TerminalRefusalReason: protocol.TerminalRefusalReasonValues,
  Theme: protocol.ThemeValues,
  ThinkingMode: protocol.ThinkingModeValues,
  TopicKind: protocol.TopicKindValues,
  ViewRefusalReason: protocol.ViewRefusalReasonValues,
};

describe("the fixed lists", () => {
  it("are the same in the generated arrays as in the Go lists", () => {
    // Adding a list in Go changes this golden file, which fails here until it is added above.
    expect(ARRAYS).toEqual(golden("enums"));
  });
});

describe("the wire rules", () => {
  it("keep the subprotocol names and the replay size the docs give", () => {
    expect(protocol.WebSocketSubprotocol).toBe("marshal.v1");
    expect(protocol.BearerSubprotocolPrefix).toBe("bearer.");
    expect(protocol.ReplayBufferSize).toBe(2000);
    expect(protocol.HomeTopic).toBe("home");
  });

  it("write every timestamp in every golden file as UTC with milliseconds", () => {
    const looksLikeATime = /^\d{4}-\d{2}-\d{2}T/;
    const wireForm = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$/;
    const times: string[] = [];
    const walk = (value: unknown): void => {
      if (typeof value === "string" && looksLikeATime.test(value)) times.push(value);
      else if (Array.isArray(value)) for (const item of value) walk(item);
      else if (value && typeof value === "object")
        for (const item of Object.values(value)) walk(item);
    };
    for (const file of readdirSync(GOLDEN_DIR).filter((name) => name.endsWith(".json"))) {
      walk(JSON.parse(readFileSync(`${GOLDEN_DIR}${file}`, "utf8")));
    }
    expect(times.length).toBeGreaterThan(0);
    for (const time of times) expect(time).toMatch(wireForm);
  });
});

File: daemon/internal/protocol/enums_test.go
package protocol_test

import (
	"testing"

	"github.com/khanblair/marshal/daemon/internal/protocol"
	"github.com/khanblair/marshal/daemon/internal/testutil"
)

// names turns a list of string-typed values into plain strings.
func names[T ~string](values []T) []string {
	out := make([]string, len(values))
	for i, v := range values {
		out[i] = string(v)
	}
	return out
}

// allEnums lists every fixed list of the protocol by its Go type name. The TypeScript tests
// compare the generated arrays with the golden file made from it.
func allEnums() map[string][]string {
	return map[string][]string{
		"CardState":             names(protocol.CardStateValues()),
		"PermissionMode":        names(protocol.PermissionModeValues()),
		"ThinkingMode":          names(protocol.ThinkingModeValues()),
		"AgentKind":             names(protocol.AgentKindValues()),
		"AgentStatus":           names(protocol.AgentStatusValues()),
		"SessionState":          names(protocol.SessionStateValues()),
		"FeedKind":              names(protocol.FeedKindValues()),
		"NoticeKind":            names(protocol.NoticeKindValues()),
		"ActivityKind":          names(protocol.ActivityKindValues()),
		"ActivityState":         names(protocol.ActivityStateValues()),
		"ChatMessageKind":       names(protocol.ChatMessageKindValues()),
		"ChatPlanState":         names(protocol.ChatPlanStateValues()),
		"ChatApprovalState":     names(protocol.ChatApprovalStateValues()),
		"CIState":               names(protocol.CIStateValues()),
		"CardViewMode":          names(protocol.CardViewModeValues()),
		"ProjectSource":         names(protocol.ProjectSourceValues()),
		"ErrorCode":             names(protocol.ErrorCodeValues()),
		"TopicKind":             names(protocol.TopicKindValues()),
		"EventType":             names(protocol.EventTypeValues()),
		"ResyncReason":          names(protocol.ResyncReasonValues()),
		"FrameType":             names(protocol.FrameTypeValues()),
		"DeviceKind":            names(protocol.DeviceKindValues()),
		"NeedsReasonKind":       names(protocol.NeedsReasonKindValues()),
		"LabelColor":            names(protocol.LabelColorValues()),
		"MoveRefusalReason":     names(protocol.MoveRefusalReasonValues()),
		"ChatTargetKind":        names(protocol.ChatTargetKindValues()),
		"ChatRefusalReason":     names(protocol.ChatRefusalReasonValues()),
		"HoldRefusalReason":     names(protocol.HoldRefusalReasonValues()),
		"ViewRefusalReason":     names(protocol.ViewRefusalReasonValues()),
		"TerminalKey":           names(protocol.TerminalKeyValues()),
		"TerminalRefusalReason": names(protocol.TerminalRefusalReasonValues()),
		"DiffFileStatus":        names(protocol.DiffFileStatusValues()),
		"DiffLineKind":          names(protocol.DiffLineKindValues()),
		"Theme":                 names(protocol.ThemeValues()),
		"ProjectView":           names(protocol.ProjectViewValues()),
		"Swimlane":              names(protocol.SwimlaneValues()),
		"FilterKey":             names(protocol.FilterKeyValues()),
		"SortDirection":         names(protocol.SortDirectionValues()),
		"ProgressStatus":        names(protocol.ProgressStatusValues()),
	}
}

func TestEnumsGolden(t *testing.T) {
	testutil.Golden(t, "enums", allEnums())
}

// The words below are the ones the app's screens use (apps/web/src/mock), as wire values.
func TestEnumsCarryTheWordsTheScreensUse(t *testing.T) {
	want := map[string][]string{
		"CardState":      {"backlog", "planning", "working", "needs", "review", "ready", "merging", "done"},
		"PermissionMode": {"ask", "auto-edits", "plan", "full-auto", "bypass"},
		"ThinkingMode":   {"low", "medium", "high", "extra-high"},
		"AgentKind":      {"claude", "gemini", "codex", "builtin"},
		"AgentStatus":    {"supported", "untested", "missing"},
		"SessionState":   {"starting", "awake", "working", "waiting-approval", "sleep-warning", "asleep", "waking", "stopped"},
		"FeedKind":       {"brief", "merge", "schedule", "approval", "plan", "ci", "tool"},
		"NoticeKind":     {"sleep", "ci-main", "cost", "plan", "ci"},
		"ActivityKind":   {"file", "command", "test", "tool", "approval"},
		"CIState":        {"queued", "running", "passed", "failed", "cancelled"},
		"CardViewMode":   {"chat", "terminal"},
	}
	got := allEnums()
	for name, values := range want {
		if !equalStrings(got[name], values) {
			t.Errorf("%s = %v, want %v", name, got[name], values)
		}
	}
}

func equalStrings(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func TestValid(t *testing.T) {
	if !protocol.CardStateNeeds.Valid() || protocol.CardState("needs-you").Valid() {
		t.Error("CardState.Valid accepts the wrong words")
	}
	if !protocol.PermissionModeBypass.Valid() || protocol.PermissionMode("Bypass permissions").Valid() {
		t.Error("PermissionMode.Valid must accept wire values only, not labels")
	}
	if !protocol.ThinkingModeExtraHigh.Valid() || protocol.ThinkingMode("Extra high").Valid() {
		t.Error("ThinkingMode.Valid must accept wire values only, not labels")
	}
	if !protocol.AgentKindClaude.Valid() || protocol.AgentKind("").Valid() {
		t.Error("AgentKind.Valid accepts the wrong words")
	}
}
